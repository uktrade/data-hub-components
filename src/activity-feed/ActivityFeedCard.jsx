import React from 'react'
import { find } from 'lodash'

import PropTypes from 'prop-types'
import Cards from './cards'

export default class ActivityFeedCard extends React.PureComponent {
  static propTypes = {
    activity: PropTypes.object.isRequired,
    showDetails: PropTypes.bool,
  }

  static defaultProps = {
    showDetails: false,
  }

  constructor(props) {
    super(props)
    const { activity } = this.props
    this.Card = find(Cards, c => c.canRender(activity))
  }

  render() {
    const { activity, showDetails } = this.props
    const { Card } = this

    return Card ? <Card activity={activity} showDetails={showDetails} /> : null
  }
}
