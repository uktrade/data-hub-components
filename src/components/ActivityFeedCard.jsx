import React from 'react'
import PropTypes from 'prop-types'
import { Link, InsetText, Table, Paragraph, H3 } from 'govuk-react'
import { filter, includes, get, map } from 'lodash'
import moment from 'moment'

const getPeople = (activity, personSubType) => {
  return map(filter(activity['object']['attributedTo'], ({ type }) => {
      return includes(type, `dit:${personSubType}`)
    }), ({ id, name, 'dit:emailAddress': emailAddress }) => {
      return {
        id,
        name,
        emailAddress,
      }
    })
}

export default class ActivityFeedCard extends React.Component {
  static propTypes = {
    activity: PropTypes.object,
  }

  constructor(props) {
    super(props)

    this.state = {
      showMoreContent: false,
    }

    this.onMoreContentClick = this.onMoreContentClick.bind(this)
  }

  onMoreContentClick(event) {
    event.preventDefault()

    this.setState({ showMoreContent: !this.state.showMoreContent })
  }

  render() {
    const { activity } = this.props
    const published = moment(activity.published).fromNow()
    const advisers = getPeople(activity, 'Adviser')
    const subject = get(activity, 'object.dit:subject')
    const service = get(activity, 'object.dit:service.name')
    const url = get(activity, 'object.url')

    return (
      <div style={{ border: '1px solid #c0c0c0', padding: '10px' }}>
        <div style={{ display: 'flex', flexFlow: 'row wrap' }}>
          <div style={{ flexGrow: 1, marginBottom: '10px' }}>
            <Paragraph>An interaction with the company took place</Paragraph>
          </div>
          <div style={{ flexGrow: 1, marginBottom: '10px', textAlign: 'right' }}>
            <Paragraph>{published}</Paragraph>
          </div>
          <div style={{ width: '100%', marginBottom: '10px' }}>
            <H3>{subject}</H3>
          </div>
        </div>
        <Link href="#" onClick={this.onMoreContentClick}>Who was involved</Link>
        {
          this.state.showMoreContent ? <div>
            <InsetText>
              <Table>
                <Table.Row>
                  <Table.CellHeader>Advisers</Table.CellHeader>
                  <Table.Cell>
                    {advisers.map(({ id, name, emailAddress }) => {
                      return <React.Fragment key={id}>
                          <span>
                            {name} <Link href={"mailto:" + emailAddress}>{emailAddress}</Link>
                          </span>
                      </React.Fragment>
                    })}
                  </Table.Cell>
                </Table.Row>
                <Table.Row>
                  <Table.CellHeader>DIT team</Table.CellHeader>
                  <Table.Cell>
                    team
                  </Table.Cell>
                </Table.Row>
                <Table.Row>
                  <Table.CellHeader>Services</Table.CellHeader>
                  <Table.Cell>{service}</Table.Cell>
                </Table.Row>
              </Table>
              <Link href={url}>Go to the interaction detail page</Link>
            </InsetText>
          </div> : null
        }
      </div>
    )
  }
}
