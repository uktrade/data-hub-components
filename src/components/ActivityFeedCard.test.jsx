import React from 'react'
import renderer from 'react-test-renderer'

import ActivityFeedCard from './ActivityFeedCard'
import interactionActivityFixture from '../../fixtures/activity_feed/interactions/interaction'

it('handles incomplete data', () => {
  const tree = renderer
    .create(<ActivityFeedCard activity={{}}/>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders interaction activity', () => {
  const tree = renderer
    .create(<ActivityFeedCard activity={interactionActivityFixture}/>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
