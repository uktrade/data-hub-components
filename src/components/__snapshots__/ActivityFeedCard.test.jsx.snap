// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles incomplete data 1`] = `null`;

exports[`renders interaction activity 1`] = `
<div
  style={
    Object {
      "border": "1px solid #c0c0c0",
      "padding": "10px",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexFlow": "row wrap",
      }
    }
  >
    <div
      style={
        Object {
          "flexGrow": 1,
          "marginBottom": "10px",
        }
      }
    >
      <div
        className="sc-jAaTju fBEFVB"
      >
        <p>
          An interaction with the company took place
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "flexGrow": 1,
          "marginBottom": "10px",
          "textAlign": "right",
        }
      }
    >
      <div
        className="sc-jAaTju fBEFVB"
      >
        <p>
          14 days ago
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "marginBottom": "10px",
          "width": "100%",
        }
      }
    >
      <h3
        className="sc-cMljjf kCLKvB"
        size="MEDIUM"
        style={
          Object {
            "color": "#005ea5",
            "fontWeight": "normal",
          }
        }
      >
        Invited to Technology Roundtable
      </h3>
    </div>
  </div>
  <details
    className="sc-kEYyzF ihcnje"
    open={false}
    style={
      Object {
        "fontSize": "100%",
        "marginBottom": 0,
      }
    }
  >
    <summary
      className="sc-kkGfuU cCdZFh"
    >
      <span
        className="sc-iAyFgw jRkIxn"
      >
        Who was involved
      </span>
    </summary>
    <div
      className="sc-hSdWYo dfimje"
    >
      <table
        className="sc-bwCtUz iVExri"
      >
        <tbody
          className="sc-eXEjpC fuVqJC"
        >
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              Advisers
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              <span>
                Mario Bros
                 
                <a
                  className="sc-bwzfXH VXMCV"
                  href="mailto:mariobros@dit.com"
                  muted={false}
                >
                  mariobros@dit.com
                </a>
              </span>
            </td>
          </tr>
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              DIT team
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              team
            </td>
          </tr>
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              Services
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              Account Management
            </td>
          </tr>
        </tbody>
      </table>
      <a
        className="sc-bwzfXH VXMCV"
        href="https://example.com/companies/a4ead59f-dfc8-e511-a5e6-e4115bead28a/interactions/a066cc49-b679-4cbe-93c5-3ceba2f067ec"
        muted={false}
      >
        Go to the interaction detail page
      </a>
    </div>
  </details>
</div>
`;
